tonatona-email/src/Tonatona/Email.hs

```diff
@@ -0,0 +1,22 @@
+module Tonatona.Email where
+
+import Network.Mail.Mime (Mail)
+
+import Tonatona (TonaM, readerShared)
+
+class TonaEmailShared shared where
+  shared :: shared -> Shared
+
+data Shared = Shared
+  { sendEmail :: forall conf shared. Mail -> TonaM conf shared ()
+  }
+
```

Comment:
In Tonatona architecture, I intend the `Shared` type is created on application start up and **shared** among all `Tonatona.{the module}.run`.
For example `Tonatona.Db.Shared` must be **shared** among all `Tonatona.Db.*.run`, and should not be created on each `run` separately.

From that point of view, this PR's usage of `Shared` seems to be invalid. It is not **shared data** but just a **common interface**.
One valid approach is to declare `sendEmail` function on each `Tonatona.Email.*` module such as `Tonatona.Email.Sendmail.sendEmail`, `Tonatona.Email.SES.sendEmail`, e.t.c.
The other approach, if you prefer to prepare common interface, is declaring a type class for common email interface as follows.

```
-- In `Tonatona.Email`
class TonaEmailInterface shared where
  send :: forall conf shared. shared -> Mail -> TonaM conf shared ()

-- In `Tonatona.Email.Sendmail`
data Shared = Shared ()
init :: IO Shared
init = pure $ Shared ()
class TonaSendmailShared shared where
  shared :: shared -> Shared
instance (TonaSendmailShared shared) => TonaEmailInterface shared where
  send _shared = liftIO . renderSendMail

-- In `Tonatona.Email.Ses`
data Shared = Shared
    { token :: SesToken
    }
init :: Config -> IO Shared
init conf = getTokenForSes conf
class TonaSesShared shared where
  shared :: shared -> Shared
instance (TonaSesShared shared) => TonaEmailInterface shared where
  send shared' mail = undefined
    where
      Shared {token} = shared shared'
```

Is it possible?

Reply:
I've changed the code around based on this comment.

The `Tonatona.Email.Sendmail` module actually does not need any configuration options or state to be used.  I've refactored is so that the `send` function is very simple:

```haskell
import Network.Mail.Mime (Mail, renderSendMail)

send :: Mail -> TonaM conf shared ()
send = liftIO . renderSendMail
```

This way the `send` function can be used without having to declare an instance of a class like `TonaEmailShared` or `TonaEmailConfig`.

How does this look?

If we add a `tonatona-email-ses` package, we will probably have to add a `TonaEmailSesConfig` class, since the it looks like the sending function in the `mime-mail-ses` package has configuration options specific to SES.

Reply:
Also, the implementation of the `tonatona-db-*` packages is similar to above.

There are configuration options to specify the connection string to the database and how many simultaneous connections to create.  The following is from the `Tonatona.Db` module:

```haskell
newtype DbConnStr = DbConnStr
  { unDbConnStr :: ByteString
  } deriving newtype (Eq, IsString, Read, Show, Var)

newtype DbConnNum = DbConnNum { unDbConnNum :: Int }
  deriving newtype (Eq, Num, Read, Show, Var)

data Config = Config
  { dbConnString :: DbConnStr
  , dbConnNum :: DbConnNum
  }
  deriving (Show)

class TonaDbConfig config where
  config :: config -> Config
```

Then, the `Shared` data type doesn't directly hold connection data, but a function that can be run to do a query.  (Internally, this function will have a reference to the `ConnectionPool`, but that is not evident from its type):

```haskell
class TonaDbShared backend shared | shared -> backend where
  shared :: shared -> Shared backend

data Shared backend = Shared
  { runDb :: forall conf shared a. ReaderT backend (TonaM conf shared) a -> TonaM conf shared a
  }
```

`Tonatona.Db` then provides a way to run any query as long as there is an instance of `TonaDbShared`:

```haskell
run :: (TonaDbShared backend shared) => ReaderT backend (TonaM conf shared) a -> TonaM conf shared a
run query = do
  f <- reader (runDb . shared . snd)
  f query
```

---------------------------------------------------------------------------

This is somewhat of an implementation detail, but the `Tonatona.Db.Sql` module defines a specialization of the `TonaDbShared` class when the `backend` parameter is `SqlBackend`:

```haskell
class TonaDbSqlShared shared where
  shared :: shared -> Shared

instance {-# OVERLAPPABLE #-} TonaDbSqlShared shared => TonaDbShared SqlBackend shared where
  shared = Tonatona.Db.Sql.shared
```

---------------------------------------------------------------------------

Finally, the `Tonatona.Db.Sql.Persistent` module gives a way to create the `Shared` datatype that was defined in `Tonatona.Db`:

```haskell
genConnectionPool :: Config -> (Loc -> LogSource -> LogLevel -> LogStr -> IO ()) -> IO (Pool SqlBackend)

runPostgres :: MonadUnliftIO m => Pool SqlBackend -> ReaderT SqlBackend m a -> m a

init :: forall config. TonaDbConfig config => config -> (Loc -> LogSource -> LogLevel -> LogStr -> IO ()) -> IO Shared
init conf logger = do
  pool <- liftIO $ genConnectionPool (TonaDb.config conf) logger
  pure $ mkShared $ \query -> runPostgres pool query
```

---------------------------------------------------------------------------

This is implemented similarly to the way that `tonatona-email` was originally implemented.  Do you think this should be changed as well?

Reply:
> Do you think this should be changed as well?

Yes. `Shared` should only have pure values as its fields.
So I recommend to change `TonaDb.Shared` to bellow.

```haskell
data Shared backend = Shared
  { connectionPool :: ConnectionPool
  }
```

Of course, if `ConnectionPool` is not actually needed to any DB connection, only `Tonatona.Db.*` modules should expose their own `Shared` data instead of exposing common `Tonatona.Db.Shared`.

What do you think about it?

Reply:
> This way the send function can be used without having to declare an instance of a class like TonaEmailShared or TonaEmailConfig.
>
> How does this look?

It seems it is what I said by following:

> One valid approach is to declare sendEmail function on each Tonatona.Email.* module such as Tonatona.Email.Sendmail.sendEmail, Tonatona.Email.SES.sendEmail, e.t.c.

Make sure that this approach is simple but there are no warranty that `Tonatona.Email.*` modules are replaceable only by changing module name as follows.

```haskell
import Tonatona.Email.Sendmail as TonaEmail

...
... do $
    TonaEmail.send
...

-- Ideally, if we need to use SES to send email, only we have to do should be changing `import` statement.
import Tonatona.Email.SES as TonaEmail

...
... do $
    TonaEmail.send
...

-- When `Tonatona.Email.SES` does not expose `send` function, this edited code cannot be compiled.
-- The other approach that prepare common interface is to guarantee this.
```

Reply:
> Make sure that this approach is simple but there are no warranty that `Tonatona.Email.*` modules are replaceable only by changing module name as follows.

I think in the `Tonatona.Email.*` modules, they would be replacable my changing the module name.  A longer example would look something like the following.

Here is the current `Tonatona.Email.Sendmail` module:

```haskell
module Tonatona.Email.Sendmail where

send :: Mail -> TonaM conf shared ()
send mail = ...
```

Here is basically what the `Tonatona.Email.SES` module would look like:

```haskell
module Tonatona.Email.SES where

send :: TonaEmailSESShared shared => Mail -> TonaM conf shared ()
send mail = ...

data Shared = Shared SESAuthData

class TonaEmailSESShared shared of
  shared :: shared -> Shared

data Config = Config SESAuthKey SESAuthSecret

instance FromEnv Config where ...

class TonaEmailSESConfig config where
  config :: config -> Config

init :: forall config. TonaEmailSESConfig config => config -> IO Shared
```

If the end user started off using `Tonatona.Email.Sendmail`, and then decided to start using `Tonatona.Email.SES`, they would have to switch the import from `import Tonatona.Email.Sendmail as TonaEmail` to `import Tonatona.Email.SES as TonaEmail`.  Then, they would also have to declare instances of `TonaEmailSESShared` and `TonaEmailSESConfig`.

If you wanted to go with a design for Tonatona where the user would have to only change the very minimal things when using a different library (for example, `tonatona-email-sendmail` vs `tonatona-email-ses`), then maybe we should make sure that `tonatona-email-sendmail` also exports the same `Shared` and `Config` data types, as well as `TonaEmailShared` and `TonaEmailConfig` classes.

Does this sound like a good idea?

Reply:
> I think in the Tonatona.Email.* modules, they would be replacable my changing the module name.

You are right. What I wanted to say is "This can be available, but it relies on whether library author of `Tonatona.Email.*` expose function  `send :: Mail -> TonaM conf shared ()`".

Though it is one good approach to **ask** library authors to export same functions, there are another approach to declare interfaces as a type class in `Tonatona.Email` module as follows.

```
class TonaEmailInterface shared where
  send :: forall conf shared. shared -> Mail -> TonaM conf shared ()
```

I don't know which approach is better for tonatona architecture. What do you think?